File: MMORPGServer/Proto/CreateLuaProto/GameLevel_ResurgenceReturnProto.lua
--服务器返回复活消息
GameLevel_ResurgenceReturnProto = { ProtoCode = 12008, IsSuccess = false, MsgCode = 0 }

--这句是重定义元表的索引，就是说有了这句，这个才是一个类
GameLevel_ResurgenceReturnProto.__index = GameLevel_ResurgenceReturnProto;

function GameLevel_ResurgenceReturnProto.New()
    local self = { }; --初始化self
    setmetatable(self, GameLevel_ResurgenceReturnProto); --将self的元表设定为Class
    return self;
end


--发送协议
function GameLevel_ResurgenceReturnProto.SendProto(proto)

    local ms = CS.LuaHelper.Instance:CreateMemoryStream();
    ms:WriteUShort(proto.ProtoCode);

    ms:WriteBool(proto.IsSuccess);
    if(not proto.IsSuccess) then
        ms:WriteInt(MsgCode);
    end

    CS.LuaHelper.Instance:SendProto(ms:ToArray());
    ms:Dispose();
end


--解析协议
function GameLevel_ResurgenceReturnProto.GetProto(buffer)

    local proto = GameLevel_ResurgenceReturnProto.New(); --实例化一个协议对象
    local ms = CS.LuaHelper.Instance:CreateMemoryStream(buffer);

    proto.IsSuccess = ms:ReadBool();
    if(not proto.IsSuccess) then
        proto.MsgCode = ms:ReadInt();
    end

    ms:Dispose();
    return proto;
end

File: MMORPGServer/GameServerApp/GameServerApp/Proto/GameLevel/GameLevel_ResurgenceReturnProto.cs
//===================================================
//作    者：边涯  http://www.u3dol.com  QQ群：87481002
//创建时间：2020-03-15 10:58:27
//备    注：
//===================================================
using System.Collections;
using System.Collections.Generic;
using System;

/// <summary>
/// 服务器返回复活消息
/// </summary>
public struct GameLevel_ResurgenceReturnProto : IProto
{
    public ushort ProtoCode { get { return 12008; } }

    public bool IsSuccess; //是否成功
    public int MsgCode; //消息码

    public byte[] ToArray()
    {
        using (MMO_MemoryStream ms = new MMO_MemoryStream())
        {
            ms.WriteUShort(ProtoCode);
            ms.WriteBool(IsSuccess);
            if(!IsSuccess)
            {
                ms.WriteInt(MsgCode);
            }
            return ms.ToArray();
        }
    }

    public static GameLevel_ResurgenceReturnProto GetProto(byte[] buffer)
    {
        GameLevel_ResurgenceReturnProto proto = new GameLevel_ResurgenceReturnProto();
        using (MMO_MemoryStream ms = new MMO_MemoryStream(buffer))
        {
            proto.IsSuccess = ms.ReadBool();
            if(!proto.IsSuccess)
            {
                proto.MsgCode = ms.ReadInt();
            }
        }
        return proto;
    }
}